concordium-consensus/src/Concordium/KonsensusV1/TreeState/LowLevel/LMDB.hs

```diff
@@ -504,55 +648,250 @@ instance
-                                    ++ show (getHash sb :: BlockHash)
-                                    ++ " at height "
-                                    ++ show h
-                            asWriteTransaction $ \dbh txn -> do
-                                -- If we haven't already, remove the latest finalization entry
-                                unless (ctr == 0) . void $
-                                    deleteRecord
-                                        txn
-                                        (dbh ^. latestFinalizationEntryStore)
-                                        CSKLatestFinalizationEntry
-                                -- Remove the block
-                                _ <- deleteRecord txn (dbh ^. blockStore) h
-                                _ <- deleteRecord txn (dbh ^. blockHashIndex) (getHash sb)
-                                -- Remove the block transactions
-                                forM_ (blockTransactions sb) $ \tx ->
-                                    deleteRecord txn (dbh ^. transactionStatusStore) (getHash tx)
-                            roll $! ctr + 1
-        getLast = asReadTransaction $ \dbh txn ->
-            withCursor txn (dbh ^. blockStore) (getCursor CursorLast)
-
 -- |Initialise the low-level database by writing out the genesis block and initial round status.
 initialiseLowLevelDB ::
     (MonadIO m, MonadReader r m, HasDatabaseHandlers r pv, MonadLogger m, IsProtocolVersion pv) =>
     -- |Genesis block.
     StoredBlock pv ->
-    -- |Initial round status.
+    -- |Initial persistent round status.
     PersistentRoundStatus ->
     DiskLLDBM pv m ()
 initialiseLowLevelDB genesisBlock roundStatus = asWriteTransaction $ \dbh txn -> do
-    storeReplaceRecord txn (dbh ^. blockStore) 0 genesisBlock
-    storeReplaceRecord txn (dbh ^. blockHashIndex) (getHash genesisBlock) 0
+    storeReplaceRecord txn (dbh ^. blockStore) (getHash genesisBlock) genesisBlock
+    storeReplaceRecord txn (dbh ^. finalizedBlockIndex) 0 (getHash genesisBlock)
     storeReplaceRecord txn (dbh ^. roundStatusStore) CSKRoundStatus roundStatus
+
+-- |Remove certified and finalized blocks from the database whose states cannot be loaded.
+-- This can throw an exception if the database recovery was not possible.
+--
+-- This uses the following assumptions:
+--   * If a block's state can be loaded, then so can the state of its parent, or any block
+--     with a lesser 'BlockStateRef'.
+--   * The database invariants hold.
+--   * No other concurrent accesses are made to the database.
+--
+-- The updates to the database are performed in the following write transactions:
+--   * Removing a single certified block.
+--   * Removing all certified blocks.
+--   * Removing the last finalized block and implicitly finalized blocks, rolling back the
+--     latest finalization entry to the prior explicitly finalized block (or removing it if
+--     it would be for the genesis block).
+rollBackBlocksUntil ::
```

Comment:
I believe we should also clear the "last timeout certificate" when rolling back blocks.

Reply:
We won't clear it, but if it's not coherent with the new highest certified block we will warn instead of crashing.